src/xsdba/utils.py

```diff
@@ -0,0 +1,1238 @@
+        # Compute new error
+        signs[orient] = np.linalg.norm(P - ((orient * R) @ Hinv) @ P)
+    return np.array(min(signs, key=lambda o: signs[o]))
+
+
+def best_pc_orientation_full(
+    R: np.ndarray,
+    Hinv: np.ndarray,
+    Rmean: np.ndarray,
+    Hmean: np.ndarray,
+    hist: np.ndarray,
+) -> np.ndarray:
+    """Return best orientation vector for `A` according to the method of :cite:t:`alavoine_distinct_2022`.
+
+    Eigenvectors returned by `pc_matrix` do not have a defined orientation.
+    Given an inverse transform `Hinv`, a transform `R`, the actual and target origins `Hmean` and `Rmean` and the matrix
+    of training observations `hist`, this computes a scenario for all possible orientations and return the orientation
+    that maximizes the Spearman correlation coefficient of all variables. The correlation is computed for each variable
+    individually, then averaged.
+
+    This trick is explained in :cite:t:`alavoine_distinct_2022`.
+    See docstring of :py:func:`sdba.adjustment.PrincipalComponentAdjustment`.
+
+    Parameters
+    ----------
+    R : np.ndarray
+      MxM Matrix defining the final transformation.
+    Hinv : np.ndarray
+      MxM Matrix defining the (inverse) first transformation.
+    Rmean : np.ndarray
+      M vector defining the target distribution center point.
+    Hmean : np.ndarray
+      M vector defining the original distribution center point.
+    hist : np.ndarray
+      MxN matrix of all training observations of the M variables/sites.
+
+    Returns
+    -------
+    np.ndarray
+      M vector of orientation correction (1 or -1).
+
+    References
+    ----------
+    :cite:cts:`alavoine_distinct_2022`
+
+    See Also
+    --------
+    sdba.adjustment.PrincipalComponentAdjustment
+    """
+    # All possible orientation vectors
+    m = R.shape[0]
+    signs = dict(itertools.zip_longest(itertools.product(*[[1, -1]] * m), [None]))
+    for orient in list(signs.keys()):
+        # Calculate scen for hist
+        scen = np.atleast_2d(Rmean).T + ((orient * R) @ Hinv) @ (
+            hist - np.atleast_2d(Hmean).T
+        )
+        # Correlation for each variable
+        corr = [spearmanr(hist[i, :], scen[i, :])[0] for i in range(hist.shape[0])]
+        # Store mean correlation
+        signs[orient] = np.mean(corr)
+    # Return orientation that maximizes the correlation
+    return np.array(max(signs, key=lambda o: signs[o]))
+
+
+def get_clusters_1d(
+    data: np.ndarray, u1: float, u2: float
+) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    """Get clusters of a 1D array.
+
+    A cluster is defined as a sequence of values larger than u2 with at least one value larger than u1.
+
+    Parameters
+    ----------
+    data : 1D ndarray
+      Values to get clusters from.
+    u1 : float
+      Extreme value threshold, at least one value in the cluster must exceed this.
+    u2 : float
+      Cluster threshold, values above this can be part of a cluster.
```

Comment:
```suggestion
    data : 1D ndarray
        Values to get clusters from.
    u1 : float
        Extreme value threshold, at least one value in the cluster must exceed this.
    u2 : float
        Cluster threshold, values above this can be part of a cluster.
```